cpp/src/arrow/device.h

```diff
@@ -29,6 +29,25 @@
 
 namespace arrow {
 
+/// \brief EXPERIMENTAL: Device type enum which matches up with C Data Device types
+enum class DeviceType : char {
```

Comment:
Is DeviceType the appropriate label for this? It seems to be more a combination of DeviceType with different allocation flavors that the drivers implement for the various device types.

Reply:
Good point, perhaps call it `DeviceAllocationType`?

Reply:
I was keeping it named "DeviceType" so that it aligns with the C ABI `ArrowDeviceType` typedef. Do you think it matters for this to be named differently?

Reply:
If you think it will cause confusion then that is fine. I found the name DeviceType misleading. I take it updating the C ABI as well is a much more involved task?

Reply:
Since we have a class `arrow::Device` which doesn't map 1-to-1 to these enum members, then I think it would be good to use another name indeed.

As for changing the C ABI: well, it wouldn't change the _ABI_, precisely :-) Also the C Device Data Interface is very new and still experimental... However, the names would probably get quite long as well.

Reply:
In that case i'll update this name to `DeviceAllocationType`